ml_perf/reference_implementation.py

```diff
@@ -0,0 +1,279 @@
+  buffer = example_buffer.ExampleBuffer(sampling_frac=1.0)
+  buffer.parallel_fill(tensorflow.gfile.Glob(pattern))
+  buffer.flush(
+      os.path.join(fsdb.golden_chunk_dir(), play_output_name + '.tfrecord.zz'))
+
+
+# Train a new model.
+def train(state, tf_records):
+  result = checked_run([
+      'python3',
+      'train.py',
+      *tf_records,
+      '--export_path={}'.format(state.train_model_path),
+  ] + py_flags(state), 'training')
+  logging.info(get_lines(result, make_slice[-8:-8]))
+
+
+# Validate the trained model against holdout games.
+def validate(state, holdout_glob):
+  result = checked_run(
+      ['python3', 'validate.py', holdout_glob] + py_flags(state),
+      'validation')
+  logging.info(get_lines(result, make_slice[-4:-3]))
+
+
+# Evaluate the trained model.
+def evaluate(state, args, name, slice):
+  sgf_dir = os.path.join(fsdb.eval_dir(), state.train_model_name)
+  result = checked_run([
+      'bazel-bin/cc/eval', '--parallel_games=100',
+      '--model={}.pb'.format(
+          state.train_model_path), '--sgf_dir={}'.format(sgf_dir)
+  ] + args, name)
+  result = get_lines(result, slice)
+  logging.info(result)
+  pattern = '{}\s+\d+\s+(\d+\.\d+)%'.format(state.train_model_name)
+  return float(re.search(pattern, result).group(1)) * 0.01
+
+
+# Evaluate trained model against previous best.
+def evaluate_model(state):
+  model_win_rate = evaluate(
+      state,
+      ['--num_readouts=100', '--model_two={}.pb'.format(state.play_model_path)] +
+      cc_flags(state), 'model evaluation', make_slice[-7:])
+  logging.info('Win rate %s vs %s: %.3f', state.train_model_name,
+               state.play_model_name, model_win_rate)
+  return model_win_rate
+
+
+# Evaluate trained model against a known target model.
+def evaluate_target(state):
+  target_win_rate = evaluate(
+      state,
+      ['--num_readouts=100', '--model_two={}'.format('ml_perf/target.pb')] +
+      cc_flags(state), 'target evaluation', make_slice[-7:])
+  logging.info('Win rate %s vs %s: %.3f', state.train_model_name,
+               state.play_model_name, target_win_rate)
+  return target_win_rate
+
+
+def rl_loop():
+  state = State()
+  bootstrap(state)
+  freeze(state, state.play_model_path)
+  selfplay(state)
+
+  while state.iter_num < 100:
+    holdout_glob = os.path.join(fsdb.holdout_dir(), '%06d-*' % state.iter_num,
+                                '*')
+    tf_records = os.path.join(fsdb.golden_chunk_dir(), '*.zz')
+    tf_records = sorted(tensorflow.gfile.Glob(tf_records), reverse=True)[:5]
+
+    state.iter_num += 1
+
+    # Train on shuffled game data of the last 5 selfplay rounds.
+    train(state, tf_records)
+    freeze(state, state.train_model_path)
+
+    # These could run in parallel.
+    validate(state, holdout_glob)
+    model_win_rate = evaluate_model(state)
+    target_win_rate = evaluate_target(state)
+
+    # This could run in parallel to the rest.
+    selfplay(state)
+
+    if model_win_rate >= 0.55:
```

Comment:
if you're only playing 100 games this should be higher but IDK if it matters

Reply:
should be 60/100 for 95% confidence that it's better.

Reply:
It's a shrug from me too, I'm leaving 0.55 for now. 
Once I've got this running on a VM with 8 v100s, we can run some experiments to determine better values